docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+
+### Comment Use of Panicking Calls :: required
+
+Use panicking methods only when the input has been explicitly checked to be correct.
+
+```
+// Attribute keys are checked for validity during data loading by ...
+let x = self.attribute_map[char_attribute.key];
+```
+
+If this check does not occur immediately before the data access (i.e. shortly before in the same function), comment clearly where it does occur.
+
+For example, if indices obtained from ICU data are to be trusted for indexed access, the data itself must have been validated at some earlier time (e.g. via a checking pass during data loading or use of a trusted hash).
+
+However, you should **never add a check purely in order to call a method which could otherwise panic**; in that situation you should always prefer to call the non-panicking equivalent and handle the Option or Result idiomatically.
+
+### Prefer Result over Option :: suggested
+
+When creating functions which can fail:
+* Use **Result** for all errors, or any cases where a user facing error message is needed.
+* Use **Option** for data accessors where "failing" always **only** implicitly means "no data available".
+  * Especially in cases where we expect the caller to have a reasonable response getting [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None).
+* Use a different enum for non-error cases with multiple return types (which can't use Option).
+
+# Advanced Features
+
+## Operator Overloading
+
+### No clever operators :: required
+
+Other than the comparison operators defined by [**Eq**](https://doc.rust-lang.org/std/cmp/trait.Eq.html), [**Ord**](https://doc.rust-lang.org/std/cmp/trait.Ord.html) etc. there is no obvious benefit to overloading other operators (e.g. overloading `*` to do something clever via the [Mul](https://doc.rust-lang.org/std/ops/trait.Mul.html) trait). This could be relaxed if standard Rust idioms for using particular operator overloads exists, and is well understood in the Rust community.
+
+See also [Operators and Symbols](https://doc.rust-lang.org/book/appendix-02-operators.html).
+
+## Binding Traits to Inbuilt Types
+
+Rust has neat ways to provide users with simple type conversion. Imagine if you have a function taking (reference to) a semantic type:
+
+```
+fn use_locale_id(id: &LocaleId) { … }
+```
+
+But the caller could just write:
+
+```
+use_locale_id(&"en_GB".into());
+```
+
+Rust's lets you bind traits to existing system type (e.g. `str`) for use within a module. And importantly, it lets you expose a series to trait bindings that other people can opt into if they want.
+By implementing the generic [`From<&str>`](https://doc.rust-lang.org/std/convert/trait.From.html) trait on `LocaleId` to convert from a string to a locale ID, we also get the [`Into<Foo>`](https://doc.rust-lang.org/std/convert/trait.Into.html) trait implied on `&str` for free.
+
+```
+impl From<&str> for LocaleId {
```

Comment:
IIUC, at least one of the source or target type must be a type you own.  Can't add a From to a type in a different crate/module?